fairscale/optim/grad_scaler.py

```diff
@@ -3,31 +3,74 @@
-from typing import Any, Dict
+from collections import abc, defaultdict
+from enum import Enum
+from typing import Any, Dict, List, Optional, Union
+import warnings
 
 import torch
+from torch.cuda import FloatTensor  # type: ignore
 from torch.cuda.amp import GradScaler as TorchGradScaler
+from torch.cuda.amp.common import amp_definitely_not_available
 import torch.distributed as dist
 from torch.optim import Optimizer
+from torch.optim.sgd import SGD
 
-from .oss import OSS
+from fairscale.utils import torch_version
+
+
+class _GeneralMultiDeviceReplicator(object):
+    """
+    Lazily serves copies of a tensor to requested devices.  Copies are cached per-device.
+    This class adds the cpu option to the _MultiDeviceReplicator class in PyTorch grad_scaler.py.
+    https://pytorch.org/docs/stable/_modules/torch/cuda/amp/grad_scaler.html#GradScaler
+    """
+
+    def __init__(self, master_tensor: torch.Tensor) -> None:
+        assert master_tensor.is_cuda or master_tensor.device.type == "xla" or master_tensor.device.type == "cpu"
+        self.master = master_tensor
+        self._per_device_tensors: Dict[torch.device, torch.Tensor] = {}
+
+    def get(self, device: torch.device) -> torch.Tensor:
+        retval = self._per_device_tensors.get(device, None)
+        if retval is None:
+            retval = self.master.to(device=device, non_blocking=True, copy=True)
+            self._per_device_tensors[device] = retval
+        return retval
+
+
+# Defines default_factory for GradScaler's _per_optimizer_states defaultdict,
+# as well as associated "enum" values.  Prefers defining these at top level because
+# - Lambdas can't be pickled, so we don't want to supply a lambda as the factory.
+# - Defining READY, UNSCALED, STEPPED and _refresh_per_optimizer_state within GradScaler
+#   causes a circular reference, which we'd rather avoid.
+class OptState(Enum):
+
+    READY = 0
+    UNSCALED = 1
+    STEPPED = 2
+
+
+def _refresh_per_optimizer_state() -> Dict:
+    return {"stage": OptState.READY, "found_inf_per_device": {}}
 
 
 class GradScaler(TorchGradScaler):
     def _unscale_grads_(
-        self, optimizer: Optimizer, inv_scale: torch.Tensor, found_inf: torch.Tensor, allow_fp16: bool
+        self,
+        optimizer: Optimizer,
+        inv_scale: torch.Tensor,
+        found_inf: torch.Tensor,
+        allow_fp16: bool,
     ) -> Dict[torch.device, torch.Tensor]:
         return super()._unscale_grads_(optimizer, inv_scale, found_inf, True)
 
 
 class ShardedGradScaler(TorchGradScaler):
     """
-    A shard-aware :class:`GradScaler<torch.cuda.amp.GradScaler>`, to be used in conjunction with
-    :class:`OSS` and :class:`ShardedOptimizer`.
-
-    Interface and usecases are not changed, more explanations can be found in the corresponding pytorch
-    documentation https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler
+    A shard aware Grad Scaler which enables loss scaling in FSDP with/without cpu_offload. This is a
```

Comment:
nit: Use of ShardedGradScaler is not specific to FSDP. It is used with any API that enables gradient sharding.

Reply:
modified as suggested.